### PR TITLE
bug 2005901: Guard controller pdb

### PR DESCRIPTION
**Pull request by ingvagabund**
Create a new guard controller which:
- create one guarding pod per each static operand (the guarding pod assume the operand has /readyz endpoint up and running)
- create a PDB with minAvailable when the topology is not SNO
- adjust PDB's minAvailable based on the number of master nodes available
- serves as a controller which recreates guarding pods there were deleted/terminated

### CODE REVIEW

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -127,14 +143,24 @@ func RunOperator(ctx context.Context, cc *controllercmd.ControllerContext) error
...
 		WithEvents(cc.EventRecorder).
 		WithInstaller([]string{"cluster-kube-controller-manager-operator", "installer"}).
 		WithPruning([]string{"cluster-kube-controller-manager-operator", "prune"}, "kube-controller-manager-pod").
 		WithRevisionedResources(operatorclient.TargetNamespace, "kube-controller-manager", deploymentConfigMaps, deploymentSecrets).
 		WithUnrevisionedCerts("kube-controller-manager-certs", CertConfigMaps, CertSecrets).
-		WithVersioning("kube-controller-manager", versionRecorder).
-		ToControllers()
+		WithVersioning("kube-controller-manager", versionRecorder)
+
+	isSNO, err := isSingleNodeTopology(ctx, configClient.ConfigV1())
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/operator/starter.go:154]: > The assumption is the topology does not changes. Once the topology is SNO, it's SNO forever. Thus, no need to query the topology inside the static pod controller dynamically.

I wouldn't assume that.  I'd rather see a lister wired up to a and either affirmatively deletes the PDB or creates it.

File: vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/guard/guard_controller.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+										node.Name,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		pod.Spec.Containers[0].ReadinessProbe.HTTPGet.Host = operands[node.Name].Status.PodIP
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply vendor/github.com/openshift/library-go/pkg/operator/staticpod/controller/guard/guard_controller.go:177]: It does

**Comment by ingvagabund** [comment]
Waiting for https://github.com/openshift/library-go/pull/1238
/hold

**Comment by ingvagabund** [comment]
Checking https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-kube-controller-manager-operator/568/pull-ci-openshift-cluster-kube-controller-manager-operator-master-e2e-aws/1471786217908473856/artifacts/e2e-aws/gather-extra/artifacts/pods/:
```
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-128-147.ec2.internal_guard.log
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-128-147.ec2.internal_guard_previous.log
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-174-19.ec2.internal_guard.log
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-174-19.ec2.internal_guard_previous.log
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-246-4.ec2.internal_guard.log
 openshift-kube-controller-manager_kube-controller-manager-guard-ip-10-0-246-4.ec2.internal_guard_previous.log
```

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by ingvagabund** [comment]
/hold
until https://github.com/openshift/library-go/pull/1268#discussion_r770769105 gets resolved through https://github.com/openshift/library-go/pull/1272

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by ravisantoshgudimetla** [comment]
This is flaking a lot.

Router seems to failing to come up.

**Comment by ravisantoshgudimetla** [comment]
/hold

as this is flaking

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**